### PR TITLE
[C23] Fix handling of alignas

### LINKED ISSUE #81472: alignas attribute not recognized in specifier-qualifier position of a member declaration
The following C program
```
#include <stdalign.h>
struct alignas_test { char c; char alignas (8) alignas_8; };
```
gives a compilation error when compiled by `clang -std=gnu2x`. The clang version is 17.0.4.

How to reproduce:
```
$ cat foo.c
#include <stdalign.h>
struct alignas_test { char c; char alignas (8) alignas_8; };
$ clang -std=gnu2x -c foo.c
foo.c:2:36: error: 'alignas' attribute cannot be applied to types
    2 | struct alignas_test { char c; char alignas (8) alignas_8; };
      |                                    ^
1 error generated.
```

The `alignas (8)` attribute is accepted when moved after the member name (i.e. right before the semicolon), but that is not where ISO C 23 wants to see it. Quoting ISO C 23 § 6.7.5 paragraph 2:
"An alignment specifier shall appear only
  - in the declaration specifiers of a declaration, or
  - in the specifier-qualifier list of a member declaration, or
  - in the type name of a compound literal.

An alignment specifier shall not be used in conjunction with
- either of the storage-class specifiers typedef or register, nor
- in a declaration of a function or bit-field."

According to the definitions of `member-declaration` and `specifier-qualifier-list` in ISO C 23 § 6.7.2.1, in the code above, `alignas (8)` is part of a `specifier-qualifier`; therefore it is valid.

### PR DESCRIPTION
**Pull request by AaronBallman**
In C++, alignas is an attribute specifier, while in C23, it's an alias of _Alignas, which is a type specifier/qualifier. This means that they parse differently in some circumstances.

Fixes https://github.com/llvm/llvm-project/issues/81472

### CODE REVIEW

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -114,6 +114,19 @@ C Language Changes
...
+
+  .. code-block:: c
+
+     struct alignas(8) /* was accepted, now rejected */ S {
+       char alignas(8) /* was rejected, now accepted */ C;
+     };
+     int i alignas(8) /* was accepted, now rejected */ ;
+
+  Fixes (`#81472 <https://github.com/llvm/llvm-project/issues/81472>`_).
+
```
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:129]: Should that be in the breaking change section?

File: clang/lib/Parse/ParseDecl.cpp [hunk 2]
```diff
@@ -3518,8 +3518,24 @@ void Parser::ParseDeclarationSpecifiers(
       DS.Finish(Actions, Policy);
       return;
 
-    case tok::l_square:
+    // alignment-specifier
+    case tok::kw__Alignas:
+      if (!getLangOpts().C11)
+        Diag(Tok, diag::ext_c11_feature) << Tok.getName();
```
Comment (cor3ntin) [inline clang/lib/Parse/ParseDecl.cpp:3524]: Preexisting but no compat warning in C11 mode?

**Review by cor3ntin — APPROVED** [review]
LGTM

File: clang/lib/Parse/ParseDecl.cpp (hunk 2 above)
Comment (AaronBallman) [reply clang/lib/Parse/ParseDecl.cpp:3524]: Yeah, we should probably add a compat warning here (and in a bunch of other places) -- perhaps a `good first issue`?

File: clang/docs/ReleaseNotes.rst (hunk 1 above)
Comment (AaronBallman) [reply clang/docs/ReleaseNotes.rst:129]: I can move the code example up there easily enough if you'd like.

**Review by cor3ntin — APPROVED** [review]
Still LGTM.
Will you be able to create the good first issue?

**Comment by AaronBallman** [comment]
> Still LGTM. Will you be able to create the good first issue?

Yes, or I'll fix the issue myself. One or the other. :-)

**Pull request merged**